zerver/lib/actions.py

```diff
@@ -3252,16 +3252,22 @@ def do_change_tos_version(user_profile: UserProfile, tos_version: str) -> None:
                                  event_type=RealmAuditLog.USER_TOS_VERSION_CHANGED,
                                  event_time=event_time)
 
-def do_regenerate_api_key(user_profile: UserProfile, acting_user: UserProfile) -> None:
-    old_api_key = user_profile.api_key
-    user_profile.api_key = generate_api_key()
-    user_profile.save(update_fields=["api_key"])
-
-    # We need to explicitly delete the old API key from our caches,
-    # because the on-save handler for flushing the UserProfile object
-    # in zerver/lib/cache.py only has access to the new API key.
-    cache_delete(user_profile_by_api_key_cache_key(old_api_key))
-
+def do_regenerate_api_key(user_profile: UserProfile, acting_user: UserProfile) -> str:
+    user_api_keys = UserAPIKey.objects.filter(user_profile=user_profile)
+    for user_api_key in user_api_keys:
+        # We need to explicitly delete the old API keys from our caches,
+        # because the on-save handler for flushing the UserProfile object
+        # in zerver/lib/cache.py only has access to the new API key.
+        cache_delete(user_profile_by_api_key_cache_key(user_api_key.api_key))
```

Comment:
@YagoGG Cache deletion was not there in your original commit. According to me I am clearing cache fine but please take a look if cache deletion or even cache setting is being done properly for multiple api keys case. I am a bit under-confident about caches in zulip

Reply:
This looks good, it correctly applies the changes in 2ba24bdd09910cf275dabe6052c5922ff989e7ef to the model with multiple API keys.